Sources/MongoSwift/BSON/BsonValue.swift

```diff
@@ -93,57 +65,36 @@ public protocol BsonValue {
     func encode(to storage: DocumentStorage, forKey key: String) throws
 
     /**
-    * Given a `bson_iter_t` known to have a next value, returns the next value in the iterator.
-    *
-    * - Parameters:
-    *   - iter: A `bson_iter_t` to read the next value from
-    *
-    * - Returns: A `BsonValue`
+    * Given a `DocumentIterator` known to have a next value of this type, 
+    * initializes the value.
     */
-    static func from(iter: inout bson_iter_t) -> BsonValue
+    init(from iter: DocumentIterator) throws
 }
 
 /// An extension of `Array` to represent the BSON array type.
 extension Array: BsonValue {
+
     public var bsonType: BsonType { return .array }
 
-    /**
-    * Given a BSON iterator where the next stored value is known to be
-    * an array, converts the data into an array. Assumes that the caller
-    * has verified the next value is an array.
-    *
-    * - Parameters:
-    *   - bson: A `bson_iter_t`
-    *
-    * - Side effects:
-    *   - bson is moved forward to the next value in the document
-    *
-    * - Returns: A `[BsonValue]` corresponding to the array
-    */
-    public static func from(iter: inout bson_iter_t) -> BsonValue {
+    public init(from iter: DocumentIterator) throws {
         var length: UInt32 = 0
         let array = UnsafeMutablePointer<UnsafePointer<UInt8>?>.allocate(capacity: 1)
         defer {
             array.deinitialize(count: 1)
             array.deallocate(capacity: 1)
         }
-        bson_iter_array(&iter, &length, array)
+        bson_iter_array(&iter.iter, &length, array)
 
         // since an array is a nested object with keys '0', '1', etc.,
         // create a new Document using the array data so we can recursively parse
         guard let arrayData = bson_new_from_data(array.pointee, Int(length)) else {
-            preconditionFailure("Failed to create a bson_t from array data")
+            throw MongoError.bsonDecodeError(message: "Failed to create a bson_t from array data")
         }
 
         let arrayDoc = Document(fromPointer: arrayData)
 
-        var i = 0
-        var result = [BsonValue]()
-        while let v = arrayDoc[String(i)] {
-            result.append(v)
-            i += 1
-        }
-        return result
+        let arr: [BsonValue?] = (0..<arrayDoc.count).map { arrayDoc[String($0)] }
```

Comment:
While a well-formed BSON array will have indexes starting with "0", it's possible for BSON data to violate that but still be considered valid. See: https://github.com/mongodb/specifications/blob/master/source/bson-corpus/tests/array.json

A robust/tolerant driver should iterate through the BSON array type and disregard possibly invalid keys.

As an aside, how does the Swift driver currently pass the aforementioned corpus tests? Would that bypass this bit of code?

Reply:
> so basically we only check the case for languages without a native (non-BSON data) representation since we're wrapping a `bson_t`.

Good point. I didn't consider that those tests were mainly for round-tripping data to a native language type and back to BSON. In that case, we would expect the round-tripping to correctly reset the array keys starting with "0".

> what about a skipped index? what if they keys are out of order?

I think the common rule is that when reading a BSON array into a native array type, the BSON keys should be entirely disregarded. And when writing a native type back to BSON we should start with "0" and go from there.

AFAIK, the server doesn't validate BSON array keys. When accessing BSON array indices in the query language, `array.1` simply looks up the "1" key in the BSON array, which is no different from a document apart from the type designation byte. I _would_ expect different behavior when using array operators, such as `$push`, as those do attempt to work with the BSON array as a proper list/vector.

----

Let's bring the question back to the LOC I'm commenting on. Is this BSON data coming from an untrusted source with potentially bad keys? If so, should we switch to using `append()` to build the Swift array instead of replying on `map()`?

Reply:
it could be coming from a bad source, yes. I think we could just return `arrayDoc.values`, which will make an iterator and call `values` on that, which just goes through the doc sequentially, ignoring keys.